### PR TITLE
Bugfix/issue11 generating 2024.1 models

### LINKED ISSUE #11: Unable to build models crate on develop branch
So I decided to try out migrating to the new models crate, but immediately ran into problems with the generator.

For starters, the 2.1.17-SNAPSHOT version of swagger-parser seems to no longer exist:

> [ERROR] Failed to execute goal on project redfish-codegen: Could not resolve dependencies for project com.twardyece.dmtf:redfish-codegen:jar:0.3.1-SNAPSHOT: Could not find artifact io.swagger.parser.v3:swagger-parser:jar:2.1.17-SNAPSHOT

Updating to 2.1.22 gets it building again:

```
--- a/redfish-generator/pom.xml
+++ b/redfish-generator/pom.xml
@@ -26,7 +26,7 @@
     <dependency>
       <groupId>io.swagger.parser.v3</groupId>
       <artifactId>swagger-parser</artifactId>
-      <version>2.1.17-SNAPSHOT</version>
+      <version>2.1.22</version>
     </dependency>
 
     <!-- https://mvnrepository.com/artifact/commons-cli/commons-cli -->
```

Next, I am flooded with these logger messages:

> [main] INFO com.twardyece.dmtf.openapi.DocumentParser - Inline schema created as Resource_Status_Conditions_inner. To have complete control of the model name, set the `title` field or provide a mapping using addInlineSchemaNameMapping()

After this, an all lowercase string trips up the PascalCase conversion:

> Exception in thread "main" com.twardyece.dmtf.text.ICaseConvertible$CaseConversionError: String inner is not convertible to PascalCase
	at com.twardyece.dmtf.text.PascalCaseName.<init>(PascalCaseName.java:97)
	at com.twardyece.dmtf.specification.UnversionedSchemaIdentifier.<init>(UnversionedSchemaIdentifier.java:22)
	at com.twardyece.dmtf.model.mapper.UnversionedModelTypeMapper.matchesType(UnversionedModelTypeMapper.java:18)
	at com.twardyece.dmtf.model.ModelResolver.resolvePath(ModelResolver.java:76)
	at com.twardyece.dmtf.model.ModelResolver.resolveSchema(ModelResolver.java:114)
	at com.twardyece.dmtf.model.ModelResolver.resolveSchema(ModelResolver.java:117)
	at com.twardyece.dmtf.model.context.factory.StructContextFactory.toProperty(StructContextFactory.java:43)
	at com.twardyece.dmtf.model.context.factory.StructContextFactory.lambda$makeModelContext$0(StructContextFactory.java:28)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1845)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at com.twardyece.dmtf.model.context.factory.StructContextFactory.makeModelContext(StructContextFactory.java:29)
	at com.twardyece.dmtf.FileFactory.makeModelFile(FileFactory.java:40)
	at com.twardyece.dmtf.RedfishCodegen.buildModels(RedfishCodegen.java:182)
	at com.twardyece.dmtf.RedfishCodegen.generate(RedfishCodegen.java:334)
	at com.twardyece.dmtf.RedfishCodegen.main(RedfishCodegen.java:404)
make: *** [generate.mk:45: redfish-models/src/lib.rs] Error 1

I tried a naive fix for this, which is probably wrong, but seems to get things progressing a bit more:

```
--- a/redfish-generator/src/main/java/com/twardyece/dmtf/text/PascalCaseName.java
+++ b/redfish-generator/src/main/java/com/twardyece/dmtf/text/PascalCaseName.java
@@ -1,5 +1,7 @@
 package com.twardyece.dmtf.text;
 
+import org.apache.commons.lang3.StringUtils;
+
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.List;
@@ -15,7 +17,15 @@ public class PascalCaseName implements ICaseConvertible, Comparable<PascalCaseNa
         // in PascalCase, e.g. abbreviations like "PCIe", we have to intentionally handle those before attempting to
         // parse the identifier(s) as PascalCase.
         ArrayList<String> identifiers = new ArrayList<>();
-        identifiers.add(name);
+
+       // Check for the corner case where the string is entirely lower case
+        if (StringUtils.isAllLowerCase(name)) {
+            String identifier = StringUtils.capitalize(name);
+            identifiers.add(identifier);
+        } else {
+            identifiers.add(name);
+        }
+
         int discoveredLength = 0;
         while (discoveredLength < name.length()) {
             String identifier = identifiers.get(identifiers.size() - 1);
```

The next case that blows up is a case conversion for a numeric string:

> main] INFO com.twardyece.dmtf.model.context.factory.UnitContextFactory - Creating unit struct for crate::models::storage::StorageController
Exception in thread "main" com.twardyece.dmtf.text.ICaseConvertible$CaseConversionError: String 1 is not convertible to PascalCase
	at com.twardyece.dmtf.text.PascalCaseName.<init>(PascalCaseName.java:94)
	at com.twardyece.dmtf.specification.UnversionedSchemaIdentifier.<init>(UnversionedSchemaIdentifier.java:22)
	at com.twardyece.dmtf.model.mapper.UnversionedModelTypeMapper.matchesType(UnversionedModelTypeMapper.java:18)
	at com.twardyece.dmtf.model.ModelResolver.resolvePath(ModelResolver.java:76)
	at com.twardyece.dmtf.model.ModelResolver.resolveSchema(ModelResolver.java:114)
	at com.twardyece.dmtf.model.context.factory.StructContextFactory.toProperty(StructContextFactory.java:43)
	at com.twardyece.dmtf.model.context.factory.StructContextFactory.lambda$makeModelContext$0(StructContextFactory.java:28)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1845)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at com.twardyece.dmtf.model.context.factory.StructContextFactory.makeModelContext(StructContextFactory.java:29)
	at com.twardyece.dmtf.FileFactory.makeModelFile(FileFactory.java:40)
	at com.twardyece.dmtf.RedfishCodegen.buildModels(RedfishCodegen.java:182)
	at com.twardyece.dmtf.RedfishCodegen.generate(RedfishCodegen.java:334)
	at com.twardyece.dmtf.RedfishCodegen.main(RedfishCodegen.java:404)
make: *** [generate.mk:45: redfish-models/src/lib.rs] Error 1

And my attempt to work around this:

```
--- a/redfish-generator/src/main/java/com/twardyece/dmtf/text/PascalCaseName.java
+++ b/redfish-generator/src/main/java/com/twardyece/dmtf/text/PascalCaseName.java
@@ -79,7 +89,8 @@ public class PascalCaseName implements ICaseConvertible, Comparable<PascalCaseNa
 
         if (identifiers.size() > 0) {
             String identifier = identifiers.get(0);
-            if (!Abbreviation.SPECIAL_ABBREVIATIONS.containsKey(identifier) && !Character.isUpperCase(identifier.charAt(0))) {
+            char c = identifier.charAt(0);
+            if (!Abbreviation.SPECIAL_ABBREVIATIONS.containsKey(identifier) && Character.isAlphabetic(c) && !Character.isUpperCase(c)) {
                 throw new CaseConversionError("PascalCase", name);
             }
         }
```
This was clearly not the correct fix, as the matcher then ends up blowing the stack:

> [main] INFO com.twardyece.dmtf.model.context.factory.UnitContextFactory - Creating unit struct for crate::models::zone::v1_6_3::Zone
Exception in thread "main" java.lang.StackOverflowError
	at java.base/java.util.regex.Pattern$BmpCharPropertyGreedy.match(Pattern.java:4329)
	at java.base/java.util.regex.Pattern$BmpCharProperty.match(Pattern.java:3955)
	at java.base/java.util.regex.Pattern$BmpCharPropertyGreedy.match(Pattern.java:4329)
	at java.base/java.util.regex.Pattern$BmpCharProperty.match(Pattern.java:3955)
	at java.base/java.util.regex.Pattern$Start.match(Pattern.java:3608)
	at java.base/java.util.regex.Matcher.search(Matcher.java:1728)
	at java.base/java.util.regex.Matcher.find(Matcher.java:745)
	at com.twardyece.dmtf.specification.VersionedSchemaIdentifier.isVersion(VersionedSchemaIdentifier.java:43)
	at com.twardyece.dmtf.model.mapper.VersionedModelTypeMapper.matchesName(VersionedModelTypeMapper.java:38)
	at com.twardyece.dmtf.model.ModelResolver.reverseResolveIdentifier(ModelResolver.java:95)
	at com.twardyece.dmtf.policies.AdditionalModelAttributesPolicy.lambda$applyToProperties$0(AdditionalModelAttributesPolicy.java:60)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at com.twardyece.dmtf.policies.AdditionalModelAttributesPolicy.applyToProperties(AdditionalModelAttributesPolicy.java:59)
	at com.twardyece.dmtf.policies.AdditionalModelAttributesPolicy.lambda$applyToProperties$0(AdditionalModelAttributesPolicy.java:62)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at com.twardyece.dmtf.policies.AdditionalModelAttributesPolicy.applyToProperties(AdditionalModelAttributesPolicy.java:59)
	at com.twardyece.dmtf.policies.AdditionalModelAttributesPolicy.lambda$applyToProperties$0(AdditionalModelAttributesPolicy.java:62)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
        .....
make: *** [generate.mk:45: redfish-models/src/lib.rs] Error 1

If any of my fixes are correct, I can check them in and send you a PR, but I'm otherwise stuck at this point. Any ideas on how to proceed?

### PR DESCRIPTION
**Pull request by AmateurECE**
This PR will resolve #11. Model generation succeeds, but the generator is currently mis-typing struct properties in the generated code. This will be resolved by adding a new ModelTypeMapper specialization in the generator. I will remove the "Draft" status from the PR when everything is fixed, and merge.

### CODE REVIEW

File: generate.mk
Comment (pmundt) [inline generate.mk:1]: Note that 2024.2 has since been released. The Swordfish release should also be updated for the latest, which is presently v1.2.6.

Comment (AmateurECE) [reply generate.mk:1]: Thank you! I've updated the Swordfish and Redfish models to these versions, though the registries (DSP8011) are still the latest at 2024.1 for the moment.

**Comment by AmateurECE** [comment]
Sorry for the delay. I think I've _almost_ got this figured out. Turns out that a lot of plumbing types became inline schemas between 2023.1 and 2024.1, where previously there were only two OpenAPI inline schemas in the whole spec. That wasn't something we were set up to cleanly accommodate, so it required a little bit of work. I should be done soon!

**Comment by AmateurECE** [comment]
Alright, @pmundt, I believe this PR is ready to be used! The `seuss` crate currently doesn't compile and I will want to fix that quick before merging (which I will do tonight), but I don't imagine that will block you for the time being. I also implemented a workaround in 9d5e512 that will allow us to merge #14 after this one is merged.

Please let me know if you experience any problems with it!

**Comment by pmundt** [comment]
Just to follow up on this, everything looks good, and I'm able to build on the new schemas without any problems. Thanks for your efforts here!

**Pull request merged**